Increase unit test coverage of `credentials/attestation_verifier/` by 3.1%

LyudmilaKostanyan opened the pull request:
#### Summary

This PR increases `credentials/attestation_verifier/` unit test coverage from **67.9% to 71%** by adding tests for certificate and attestation validation in the partial DAC verifier.

The new tests cover the following scenarios:

* **Expired certificates** – ensuring attestation is rejected when the DAC validity period is in the past.
* **Not-yet-valid certificates** – confirming certificates that are valid only in the future are also rejected.
* **Malformed attestation elements** – validating the verifier properly detects and rejects corrupted or non-conforming attestation data.
* **Nonce mismatch** – verifying that a nonce embedded in attestation elements must match the expected nonce provided to the verifier.
* **Invalid PAA format** – exercising the failure path where the PAA certificate is not valid, ensuring the correct error is reported.

#### Related issues

Main issue: #37234

#### Testing

This PR only adds new unit tests. No changes were made to production code.

LyudmilaKostanyan committed:
Added test with valid signature

LyudmilaKostanyan committed:
Add tests for DAC certificate expiration

- Add test TestWithExpiredDACCertificate

- Add test TestWithExpiredDACCertificate

LyudmilaKostanyan committed:
Add tests for DacOnlyPartialAttestationVerifier

- Add test for malformed attestation elements: TestWithMalformedAttestationElements

- Add test for mismatched nonce handling: TestWithMismatchedNonce

LyudmilaKostanyan committed:
Clarify and improve comments

LyudmilaKostanyan committed:
Refactor tests for clarity and consistency

LyudmilaKostanyan committed:
Skip tests if platform does not support current time retrieval for DAC certificate validation

Elen777300 commented on src/credentials/tests/TestDacOnlyPartialAttestationVerifier.cpp:
```diff
@@ -48,6 +48,25 @@ static void OnAttestationInformationVerificationCallback(void * context, const D
+ * @param keypair Reference to a P256Keypair object to populate.
+ * @return CHIP_ERROR indicating success or failure of the operation.
+ */
+CHIP_ERROR LoadKeypairFromRaw(ByteSpan private_key, ByteSpan public_key, Crypto::P256Keypair & keypair)
```
Maybe we should move this helper function to P256Keypair. 
I've seen this function all over the repository.

Elen777300 commented on src/credentials/tests/TestDacOnlyPartialAttestationVerifier.cpp:
```diff
@@ -255,20 +239,268 @@ TEST_F(TestDacOnlyPartialAttestationVerifier, TestWithInvalidAttestationSignatur
 }
+
+// Test case validates that the verifier correctly rejects DAC certificates that have expired
+TEST_F(TestDacOnlyPartialAttestationVerifier, TestWithExpiredDACCertificate)
```
TestWithExpiredDACCertificate and TestWithValidInFutureDACCertificate repeat the same code except the certificates. I think we should move this to one function with arguments.

Elen777300 commented on src/credentials/tests/TestDacOnlyPartialAttestationVerifier.cpp:
```diff
@@ -255,20 +239,268 @@ TEST_F(TestDacOnlyPartialAttestationVerifier, TestWithInvalidAttestationSignatur
+    GTEST_SKIP() << "Skipping test: platform does not support current time.";
+#endif
+    // The actual contents do not represent real or meaningful data.
+    uint8_t cdData[32] = { 0x11, 0x22, 0x33, 0x44 };
```
We can move this to the structure to be a constexpr.

Elen777300 commented:
I think it would make more sense, if we moved the `LoadKeypairFromRaw` changes to a separate PR.
There are a lot of changes and they don't need to be mixed with the unit test.

LyudmilaKostanyan committed:
Refactor attestation verifier tests

- Moved cdData to a structure to avoid repetition

- Combined repeated logic in TestWithExpiredDACCertificate and TestWithValidInFutureDACCertificate into a single helper method

- Added a helper method for creating signed attestation data to reduce code duplication

LyudmilaKostanyan committed:
Moving refactoring functions out of the structure

LyudmilaKostanyan committed:
Initialize `tlvLen` in `TestDacOnlyPartialAttestationVerifier` tests to prevent uninitialized variable usage

LyudmilaKostanyan committed:
Remove custom `HazardousOperationLoadKeypairFromRaw` implementation and use the method from `P256Keypair` class

LyudmilaKostanyan committed:
Enhance comments in `TestDacOnlyPartialAttestationVerifier` tests

LyudmilaKostanyan committed:
Refactor `TestDacOnlyPartialAttestationVerifier` by moving helper functions to the class

LyudmilaKostanyan committed:
Update  test data to use dynamic sizes for nonce and signature arrays

LyudmilaKostanyan commented on src/credentials/tests/TestDacOnlyPartialAttestationVerifier.cpp:
```diff
@@ -36,37 +36,142 @@ using namespace chip::Credentials;
+    static constexpr uint8_t kTestAttestationElements[32] = { 0x01, 0x02, 0x03 };
+    static constexpr uint8_t kTestChallengeData[32]       = { 0x04, 0x05, 0x06 };
+    static constexpr uint8_t kTestSignatureData[64]       = { 0x07, 0x08, 0x09 };
+    static constexpr uint8_t kTestNonceData[32]           = { 0x0A, 0x0B, 0x0C };
```
Updated kTestNonceData to use kAttestationNonceLength and kTestSignatureData to use kP256_ECDSA_Signature_Length_Raw constants.

LyudmilaKostanyan commented on src/credentials/tests/TestDacOnlyPartialAttestationVerifier.cpp:
```diff
@@ -65,8 +77,79 @@ struct TestDacOnlyPartialAttestationVerifier : public ::testing::Test
 };
 
+// Helper method to construct TLV attestation elements structure
+static CHIP_ERROR ConstructAttestationElementsTLV(uint8_t * tlvBuffer, size_t bufferSize, const ByteSpan & cdData,
```
Moved ConstructAttestationElementsTLV and CreateSignedAttestationData to private static methods in the test fixture.

tcarmelveilleux commented on src/credentials/tests/TestDacOnlyPartialAttestationVerifier.cpp:
```diff
@@ -36,37 +37,142 @@ using namespace chip::Credentials;
+    {
+        return CHIP_ERROR_BUFFER_TOO_SMALL;
+    }
+    uint8_t toSign[kMaxToSignSize];
+    memcpy(toSign, tlvBuffer, tlvLen);
+    memcpy(toSign + tlvLen, challengeData, challengeDataLen);
```
We have BufferWriter to make these kinds of thing safe.

LyudmilaKostanyan committed:
Refactor  to use  for concatenating data

tcarmelveilleux commented on src/credentials/tests/TestDacOnlyPartialAttestationVerifier.cpp:
```diff
@@ -36,37 +38,142 @@ using namespace chip::Credentials;
+    writer.Init(tlvBuffer, bufferSize);
+    chip::TLV::TLVType outerType;
+
+    // Add attestation elements to the TLV structure
```
Add comment showing schema from the spec. The magic numbers make it harder to validate without a side-by-side source of truth.

LyudmilaKostanyan replied:
Add comment showing schema from the spec.

LyudmilaKostanyan committed:
Clear buffers, bound TLV size, add schema comment
- Zero-initialize stack signing buffers (toSign) to avoid uninitialized memory.
- Require TLV buffer size in CreateSignedAttestationData and callers; remove hardcoded 128.
- Add Attestation Elements TLV schema comment with spec reference (Matter 1.4 §11.18.4.6).

LyudmilaKostanyan commented on src/credentials/tests/TestDacOnlyPartialAttestationVerifier.cpp:
```diff
@@ -36,37 +38,142 @@ using namespace chip::Credentials;
+
+    // Concatenate TLV and challenge data for signing
+    constexpr size_t kMaxToSignSize = 128 + 64; // 128 for TLV, 64 for challengeData
+    uint8_t toSign[kMaxToSignSize];
```
Zero-initialized toSign buffers.

LyudmilaKostanyan commented on src/credentials/tests/TestDacOnlyPartialAttestationVerifier.cpp:
```diff
@@ -36,37 +38,142 @@ using namespace chip::Credentials;
+    size_t challengeDataLen)
+{
+    // Construct TLV attestation elements
+    ReturnErrorOnFailure(ConstructAttestationElementsTLV(tlvBuffer, sizeof(uint8_t) * 128, cdData, nonceData, tlvLen));
```
Accept TLV buffer size; remove magic 128.
